exchanges/gateio/gateio_websocket_request_spot.go

```diff
@@ -0,0 +1,278 @@
+)
+
+var (
+	errBatchSliceEmpty  = errors.New("batch cannot be empty")
+	errNoOrdersToCancel = errors.New("no orders to cancel")
+	errEdgeCaseIssue    = errors.New("edge case issue")
+	errChannelEmpty     = errors.New("channel cannot be empty")
+)
+
+// websocketLogin authenticates the websocket connection
+func (g *Gateio) websocketLogin(ctx context.Context, conn stream.Connection, channel string) (*WebsocketLoginResponse, error) {
+	if conn == nil {
+		return nil, fmt.Errorf("%w: %T", common.ErrNilPointer, conn)
+	}
+
+	if channel == "" {
+		return nil, errChannelEmpty
+	}
+
+	creds, err := g.GetCredentials(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	tn := time.Now().Unix()
+	msg := "api\n" + channel + "\n" + "\n" + strconv.FormatInt(tn, 10)
+	mac := hmac.New(sha512.New, []byte(creds.Secret))
+	if _, err = mac.Write([]byte(msg)); err != nil {
+		return nil, err
+	}
+	signature := hex.EncodeToString(mac.Sum(nil))
+
+	payload := WebsocketPayload{
+		RequestID: strconv.FormatInt(conn.GenerateMessageID(false), 10),
+		APIKey:    creds.Key,
+		Signature: signature,
+		Timestamp: strconv.FormatInt(tn, 10),
+	}
+
+	req := WebsocketRequest{Time: tn, Channel: channel, Event: "api", Payload: payload}
+
+	resp, err := conn.SendMessageReturnResponse(ctx, request.Unset, req.Payload.RequestID, req)
+	if err != nil {
+		return nil, err
+	}
+
+	var inbound WebsocketAPIResponse
+	if err := json.Unmarshal(resp, &inbound); err != nil {
+		return nil, err
+	}
+
+	if inbound.Header.Status != "200" {
+		var wsErr WebsocketErrors
+		if err := json.Unmarshal(inbound.Data, &wsErr.Errors); err != nil {
+			return nil, err
+		}
+		return nil, fmt.Errorf("%s: %s", wsErr.Errors.Label, wsErr.Errors.Message)
+	}
+
+	var result WebsocketLoginResponse
+	return &result, json.Unmarshal(inbound.Data, &result)
+}
+
+// WebsocketOrderPlaceSpot places an order via the websocket connection. You can
+// send multiple orders in a single request. But only for one asset route.
+func (g *Gateio) WebsocketOrderPlaceSpot(ctx context.Context, batch []WebsocketOrder) ([]WebsocketOrderResponse, error) {
+	if len(batch) == 0 {
+		return nil, errBatchSliceEmpty
+	}
+
+	for i := range batch {
+		if batch[i].Text == "" {
+			// For some reason the API requires a text field, or it will be
+			// rejected in the second response. This is a workaround.
+			batch[i].Text = "t-" + strconv.FormatInt(g.Counter.IncrementAndGet(), 10)
+		}
+		if batch[i].CurrencyPair == "" {
+			return nil, currency.ErrCurrencyPairEmpty
+		}
+		if batch[i].Side == "" {
+			return nil, order.ErrSideIsInvalid
+		}
+		if batch[i].Amount == "" {
+			return nil, errInvalidAmount
+		}
+		if batch[i].Type == "limit" && batch[i].Price == "" {
+			return nil, errInvalidPrice
+		}
+	}
+
+	if len(batch) == 1 {
+		var singleResponse WebsocketOrderResponse
+		return []WebsocketOrderResponse{singleResponse}, g.SendWebsocketRequest(ctx, "spot.order_place", asset.Spot, batch[0], &singleResponse, 2)
+	}
```

Comment:
I think it could be nuanced in a way their matching engine operates and executes batches. If you send a single order it might have less overhead (maybe not 🤷), so if you include a single order as an outbound slice there could be differences (maybe not 🤷). I haven't added batch support in a market maker yet so I have no idea. I am still thinking we should keep the current mapping to their functionality for now, even though what you suggest is much much cleaner

Reply:
Okay. There's too much guessing in there for my liking, but I'll leave it be.